File: heap/lb-11-find-median-from-stream-lc-295.py
# You are given stream of numbers and u need to find median at any point of time.
# idea is to maintain the sorted array and then find median in O(1) time.

class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.arr = []
        
    def addNum(self, num: int) -> None:
        # add number to list 
        self.append(num)
        # get the index of newly inserted element
        i = len(self.arr) - 1
        # get that element=num
        ele = self.arr[i]
        # as the array is sorted hence
        # move and place this value in this place
        # if this element is less than previos 
        # keep moving right till its location
        while i>0 and ele<self.arr[i-1]:
            self.arr[i] = self.arr[i-1]
            i-=1
        # no place element to its location
        # if already sorted then no change.
        self.arr[i] = ele
        

    def findMedian(self) -> float:
        # as the array is sorted
        # if odd then middle is median
        # else avg of two mid elements
        n = len(self.arr)
        if n&1:
            return self.arr[n//2]
        return (self.arr[n//2 - 1] + self.arr[n//2])/2


## Optimization using heap.
# Try creating 2 heaps max and min.
# at any time max heap will contain median at root is n is odd.
# and avg of root of max and min heap is median if n is even.
# | max heap | min heap | spilt the numbers into two parts.
# TODO:
class MedianFinderHeap:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.min_heap = []
        self.max_heap = []
        
    def addNum(self, num: int) -> None:
        pass
        

    def findMedian(self) -> float:
        pass

File: heap/lb-11-find-median-from-stream-lc-295.cpp
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;
class Solution {
   public:
    // Max heap | Min heap
    // min_heap <= max_heap <= 1+ min_heap
    priority_queue<int> max_heap;
    priority_queue<int, vector<int>, greater<int>> min_heap;
    void max_heap_to_min(int val) {
        // current element is smaller than max of LHS
        int max_ele = max_heap.top();
        // so it should be place in LHS but can;t be inserted as LHS will have
        // 2 more element than RHS hence remove from LHS and then insert into LHS
        max_heap.pop();
        max_heap.push(val);
        min_heap.push(max_ele);
    }
    void min_heap_to_max(int val) {
        // current element is smaller than max of LHS
        int min_ele = min_heap.top();
        // so it should be place in LHS but can;t be inserted as LHS will have
        // 2 more element than RHS hence remove from LHS and then insert into LHS
        min_heap.pop();
        min_heap.push(val);
        max_heap.push(min_ele);
    }
    void insertHeap(int &x) {
        // first element will go into max-heap
        int mxs = max_heap.size();
        int mns = min_heap.size();
        // this is the 1st element
        if (mxs == 0) {
            max_heap.push(x);
        } else if (mns == mxs) {
            // both heap has same number of elements
            // if current element is less than all the elements of right side array.
            // push in left side as left side can contain 1 more element.
            if (x < min_heap.top()) {
                max_heap.push(x);
            } else {
                // current element is greator than smllest element of RHS
                // so it should be pushed in RHS but,
                // rhs can;t contain 1 more element hence remove from RHS and push to LHS
                // take the min element of RHS > max of LHS
                min_heap_to_max(x);
            }
        } else if (mns == 0) {
            // min heap is empty
            // current element is larger than max of LHS
            // push in minheap
            if (x > max_heap.top()) {
                min_heap.push(x);
            } else {
                max_heap_to_min(x);
            }
        } else if (x >= min_heap.top()) {
            min_heap.push(x);
        } else {
            // max heap > min heap size
            if (x < max_heap.top()) {
                // current element is smaller than max of LHS
                // so it should be place in LHS but can;t be inserted as LHS will have
                // 2 more element than RHS hence remove from LHS and then insert into LHS
                max_heap_to_min(x);
            } else {
                // more than max-heap top and less than min heap top
                min_heap.push(x);
            }
        }
    }

    // Function to balance heaps.
    void balanceHeaps() {
    }

    // Function to return Median.
    double getMedian() {
        int mns = min_heap.size();
        int mxs = max_heap.size();
        return (mxs > mns) ? double(max_heap.top()) : double((max_heap.top() + min_heap.top()) / 2);
    }
};
